File: 30_days_C(modern)_HackerRank_Challenge/BSTLevel-OrderTraversal.c
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

typedef struct Node{
    struct Node* left;
    struct Node* right;
    int data;
}Node;
Node* newNode(int data){
    Node* node=(Node*)malloc(sizeof(Node));
    node->left=node->right=NULL;
    node->data=data;
    return node;
}

void levelOrder(Node* root){
    if (root == NULL) return;

    // Use an array as a simple queue
    Node* queue[500]; // Size based on constraints
    int head = 0;
    int tail = 0;

    queue[tail++] = root;

    while (head < tail) {
        Node* current = queue[head++];
        printf("%d ", current->data);

        if (current->left != NULL) {
            queue[tail++] = current->left;
        }
        if (current->right != NULL) {
            queue[tail++] = current->right;
        }
    }

}

Node* insert(Node* root,int data){
    if(root==NULL)
        return newNode(data);
    else{
        Node* cur;
        if(data<=root->data){
            cur=insert(root->left,data);
            root->left=cur;                
        }
        else{
            cur=insert(root->right,data);
            root->right=cur;
        }
        
    }
    return root;
}
int main(){
    Node* root=NULL;
    int T,data;
    scanf("%d",&T);
    while(T-->0){
        scanf("%d",&data);
        root=insert(root,data);
    }
    levelOrder(root);
    return 0;
    
}

File: 30_days_Python_HackerRank_Challenge/BSTLevel-OrderTraversal.py
#python2
import sys

class Node:
    def __init__(self,data):
        self.right=self.left=None
        self.data = data
class Solution:
    def insert(self,root,data):
        if root==None:
            return Node(data)
        else:
            if data<=root.data:
                cur=self.insert(root.left,data)
                root.left=cur
            else:
                cur=self.insert(root.right,data)
                root.right=cur
        return root

    def levelOrder(self,root):
 
        if root is None:
            return
        
        queue = [root]
        
        while len(queue) > 0:
            current = queue.pop(0)
            
            print current.data, 
            
            if current.left:
                queue.append(current.left)
            if current.right:
                queue.append(current.right)

T=int(raw_input())
myTree=Solution()
root=None
for i in range(T):
    data=int(raw_input())
    root=myTree.insert(root,data)
myTree.levelOrder(root)



#python3
import sys

class Node:
    def __init__(self,data):
        self.right=self.left=None
        self.data = data
class Solution:
    def insert(self,root,data):
        if root==None:
            return Node(data)
        else:
            if data<=root.data:
                cur=self.insert(root.left,data)
                root.left=cur
            else:
                cur=self.insert(root.right,data)
                root.right=cur
        return root

    def levelOrder(self,root):
  
        if root is None:
            return
        
        queue = [root]
        
        while len(queue) > 0:
            node = queue.pop(0)
            
            print(node.data, end=" ")
            
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

T=int(input())
myTree=Solution()
root=None
for i in range(T):
    data=int(input())
    root=myTree.insert(root,data)
myTree.levelOrder(root)
